### PR TITLE
 Improve NXcanSAS support

### LINKED ISSUE #16685: Fix issues in Mantid's NXcanSAS support
We received feedback from teh CanSAS consortium regarding the Mantid's support of NXcanSAS file format. The are several items which need to be corrected:

>both files:
* NX_class attribute must be one of the NeXus classes
 * Change existing "NX_class" to "canSAS_class"
 * provide appropriate "NX_class" attribute (see table below)
* Time & date fields should be reported in ISO-8601 format
 * strftime format: "%Y-%m-%dT%H:%M:%S"  or  "%Y-%m-%d %H:%M:%S"
* attribute Q_indices: should be int array
* SDD is should be reported as a number, not a string

>2-D file
* on the SAS intensity "I”,  add valid attribute "units", remove empty
attribute "unit"
* same for "Idev"

>The canSAS data format has a structure similar to NeXus, not identical.
To allow canSAS data to be expressed in NeXus, yet identifiable
by the canSAS standard, an additional group attribute "canSAS_class"
was introduced.  Here is the mapping for the two data files presented.

===============  ============  ==========================
group            NX_class      canSAS_class
===============  ============  ==========================
sasentry01       NXentry       SASentry
sasdata          NXdata        SASdata
sasinstrument    NXinstrument  SASinstrument
sasprocess       NXprocess     SASprocess
sastransmission  NXdata        SAStransmission_spectrum
sasdetector      NXdetector    SASdetector
sassource        NXsource      SASsource
===============  ============  ==========================

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
~~This is currently with Stephen King and the CanSAS consortium.~~
The CanSAS committee has approved of the changes.

Fixes #16685

Several issues, reported by the CanSAS consortium, were fixed:
1. Introduce canSAS_class
2. Provide NX_class with equivalent class names. This attribute lives in parallel to the canSAS_class one. Essentially this makes the file format compatible with Nexus and CanSAS
3. Set dates to ISO-8601 format
4. Change the Q_indices attribute to be a int array
5. SDD which is part of the detectors is now a floating point instead of a string
6. change `unit` to `units` 
## For testing
- The changes will be presented to the CanSAS consortium again, hence we have scientific validation for this.
- Code review. 
  Does not have to be in release notes since it is a "small" enhancement
  ---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/H5Util.cpp [hunk 1]
```diff
@@ -104,6 +108,21 @@ void writeStrAttribute(LocationType &location, const std::string &name,
   attribute.write(attrType, value);
 }
 
+template <typename NumT, typename LocationType>
+void writeNumAttribute(LocationType &location, const std::string &name,
+                    const NumT &value) {
+  static_assert(std::is_integral<NumT>::value || std::is_floating_point<NumT>::value,
+                "The writeNumAttribute function only accepts integral of floating point values.");
```
Comment (KarlPalmen) [inline Framework/DataHandling/src/H5Util.cpp]: I think you mean "or" instead of "of".

File: Framework/DataHandling/src/H5Util.cpp [hunk 2]
```diff
@@ -209,8 +228,45 @@ std::vector<OutputNumT> convertingingRead(DataSet &dataset,
 
   return result;
 }
+
+template <typename InputNumT, typename OutputNumT>
+OutputNumT convertingingRead(Attribute &attribute,
+                                          const DataType &dataType) {
```
Comment (KarlPalmen) [inline Framework/DataHandling/src/H5Util.cpp]: Is the name convertingingRead really what is intended?

**Comment by KarlPalmen** [comment]
Code review good :shipit:

**Comment by AntonPiccardoSelg** [comment]
Gatekeeper, this is not for shipping yet, as it is still with the CANSAS steering comitee. Once the steering comitee approves, I will remove the "In Progress" flag

**Comment by FedeMPouzols** [comment]
I've just edited the comments to remove the "squirrel" word, just to avoid confusion. Please remember to re-"squirrel" this once it is checked by the CANSAS committee and ready.

**Comment by AntonPiccardoSelg** [comment]
The CanSAS committee has approved the changes, hence :shipit:

**Pull request merged**